### PR TITLE
Add a runtime test for multiple versions

### PR DESCRIPTION
**Pull request by jsturtevant**
When working on #781 I noticed there wasn't a runtime test that had versions in the wit files.  This adds test for a wit file that has multiple versions for rust and csharp.

### CODE REVIEW

File: tests/runtime/versions/world.wit
Comment (silesmo) [reply tests/runtime/versions/world.wit:7]: You should be able to alias them https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#package-format. @jsturtevant

Comment (jsturtevant) [reply tests/runtime/versions/world.wit:7]: might be able to work around it with that.  

As I am looking closer, I do think these should be different names.  The types and return values are distinct, the generation of the name should be like `test_dep0_2_0_test()` to match the type

**Comment by jsturtevant** [comment]
the C# failure the following, looks like we might be missing the versioning info on the import.

```
EXEC : error : WebAssembly function imports must be unique [D:\a\wit-bindgen\wit-bindgen\target\runtime-tests\versions\csharp-foo\Foo.csproj]
  System.Exception: WebAssembly function imports must be unique
     at ILCompiler.ConfigurableWasmImportPolicy.AddWasmImport(String) + 0x117
     at ILCompiler.ConfigurableWasmImportPolicy..ctor(IReadOnlyList`1, IReadOnlyList`1) + 0x2d4
     at ILCompiler.Program.Run() + 0x161d
```

File: tests/runtime/versions/world.wit
Comment (jsturtevant) [reply tests/runtime/versions/world.wit:7]: Discussed a bit in zulip: https://bytecodealliance.zulipchat.com/#narrow/stream/327223-wit-bindgen/topic/Host.20Exports.20across.20versions

https://github.com/bytecodealliance/wasmtime/pull/7656

**Comment by jsturtevant** [comment]
This is waiting on release with https://github.com/bytecodealliance/wasmtime/pull/7656 for rust tests.  The c# should work after #791

**Comment by alexcrichton** [comment]
Oh if you'd like it's ok to use a git dependency on `wasmtime` to pull in fixes like that

**Comment by jsturtevant** [comment]
> Oh if you'd like it's ok to use a git dependency on `wasmtime` to pull in fixes like that

sounds good, I will do that and drop the c# test until we get the other PR merged

**Comment by jsturtevant** [comment]
@alexcrichton updated, I had to move a few dependencies, a couple type changes and regenerate the adapter module

**Comment by jsturtevant** [comment]
opened #802 to update the dependencies and will rebase on top of that

**Comment by jsturtevant** [comment]
> opened #802 to update the dependencies and will rebase on top of that

ends up that wasmtime 0.16.0 didn't have the required commit. I saw the date of the release which was after my commit went in but I guess it was from a branch a little further back in time.  

wasmtime is only used as a dev dependency so I've moved it and made the required changes to the API.

**Comment by jsturtevant** [comment]
@alexcrichton up to you if you want to wait for a release for this or use the revision

**Review by alexcrichton — APPROVED** [review]
Looks good to me! Mind dropping the `rev` though in the dependency? I think the `Cargo.lock` entry should suffice for that

**Comment by jsturtevant** [comment]
Without the rev it will mean we will pull the tip of the main branch.  I just tested it and this works but could cause issues down the line as the branch progresses?

**Comment by alexcrichton** [comment]
WIth a `Cargo.lock` that won't happen unless a `cargo update` is issued which by the time that happens Wasmtime 17 will be released and it should be good to use anywya

**Pull request merged**
